### DISCUSSION TITLE
SandWorm 2.0 Next Steps

### DISCUSSION THREAD

**Discussion by mariuszhermansdorfer** [mariuszhermansdorfer/SandWorm#79] [Ideas]
@philipbelesky, @BarusXXX 
I'm more or less done with any major changes to SandWorm. Will test it with Kinect for Windows on Monday to double-check, that the new functionality didn't break too many of the old things and hopefully we'd have a more or less complete plugin ready for 2.0 release.

It might be a good moment to reflect, what the next features should be. On one hand, we have a list of comments from our student workshops (#71), on the other, functionality, we'd like to add for various reasons. I'll list these in no particular order and please feel free to add to it:

- [x] 1. Cut & Fill analysis (#44)
- [ ] 2. Proper point-cloud output
- [x] 3. Interactive flowlines (#14)
- [x] 4. User-defined color gradients (#19)
- [ ] 5. QR code / RGB beacon recognition (#25)
- [ ] 6. Tutorials (#9)
- [x] 7. Automatic calibration

I could take care of `1` & `7`. Most of the code is already there in one of the older branches and could be ported with relatively low effort. I'll probably only have time in September to look into it, though.

`2` should be fairly straightforward with Tarsier as well. I vaguely recall, there was a licensing mismatch between SandWorm and Tarsier, but as long as it is permissive I don't mind. @philipbelesky would you like to take this one on?

I can envision `3` being a bunch of cellular-automata polylines, at each tick growing by one cell in the slope direction and decaying over time. Slope logic is already there, so it too should be a relatively low hanging fruit. @BarusXXX  would you take this one on?

I'm not sure about `5-6`, any volunteers?

**Comment by mariuszhermansdorfer** [comment]
`4` is done in the latest commit, see #19 for more info.

**Comment by mariuszhermansdorfer** [comment]
Another idea would be to design a hardware mount for the Kinect Azure sensor. Ideally, it would provide very fine control over pitch & roll as defined in this graph to simplify the calibration process:
![image](https://user-images.githubusercontent.com/49192999/129478978-a47db1d4-5c08-4dcf-a261-9ff59177956c.png)

This is overkill, but I really like the fine control these screw-controlled actuators provide ([source](https://www.thorlabs.com/newgrouppage9.cfm?objectgroup_id=1171&pn=PY003/M)):
![image](https://user-images.githubusercontent.com/49192999/129479092-58a23503-9f6c-4ae5-a219-877a80a933b6.png)

@BarusXXX, any good ideas?

**Comment by philipbelesky** [comment]
I'm happy to take on `#5` and `#6` as there's a good degree of overlap with some of my future plans. 

For #3 I could definitely provide the naïve / groundhog implementation (which includes the technique shown by Ander's in one of the threads). But, for this implementation to be very speedy it probably should use a cell/pixel-based approach rather than implementing gradient descent using Rhino geometry. I'd be happy to look into it down the track though.

**Comment by mariuszhermansdorfer** [comment]
Thanks @philipbelesky. I've been thinking about `#3` for a while and would like to give it a go as well. We could try a few things in separate branches and see what works best. One of the ideas I had, was creating a separate mesh for water flow. Each cell would have evaporation/infiltration/ precipitation values. On every tick we'd calculate water flowing in/out of these cells and display the resulting water surface accordingly. It has the potential of being more dynamic than just flowlines and potentially more interesting to work with.

I'll start working on it after the second week of September the earliest though, so feel free to tackle this issue if you have time before.

**Reply by philipbelesky** [reply]
Sounds good! The [model in this plugin](https://github.com/livestock3d/livestock) might be worth a look. A consideration with the waterflow options is that it can be very valuable to have some spatial control over surface roughness/infiltration, e.g. allowing users to demarcate say a paved area or a planted area within the landform model to see how that would affect runoff. That is probably tricky to build into the core sandworm component, but might justify setting up the analysis so it could also be run downstream.

**Comment by mariuszhermansdorfer** [comment]
Calibration `#7` is done in the in the [latest commit](https://github.com/mariuszhermansdorfer/SandWorm/commit/175b7c682a853c41dc1a10b53b7d9bcfe71cfc1b)

![image](https://user-images.githubusercontent.com/49192999/131036179-7774d10a-cb2a-4875-9f74-967d730f8f44.png)

**Comment by mariuszhermansdorfer** [comment]
CutFill analysis `#1` is done in the [latest commit](https://github.com/mariuszhermansdorfer/SandWorm/tree/feature/SandWorm2-0)

See #44 for more info.

![image](https://user-images.githubusercontent.com/49192999/131132317-0d954b46-a5c5-475e-832d-09b0a0c61b08.png)

**Comment by mariuszhermansdorfer** [comment]
Water flow simulation with interactive flowlines `#3` is done in the [latest commit](https://github.com/mariuszhermansdorfer/SandWorm/commit/7665ef64065bfe55187b5d5b77741bf3ab3e564a).

![image](https://user-images.githubusercontent.com/49192999/132193054-153e5d38-ce98-45b1-a07e-530223bd5ea8.png)

See #14 for more info.
